stan/math/prim/fun/gaus_interp.hpp

```diff
@@ -0,0 +1,198 @@
+#ifndef STAN_MATH_PRIM_FUN_GAUS_INTERP
+#define STAN_MATH_PRIM_FUN_GAUS_INTERP
+
+#include <cmath>
+#include <vector>
+#include <algorithm>
+#include <stan/math/prim/fun/conv_gaus_line.hpp>
+
+using namespace std;
+
+namespace stan {
+namespace math {
+
+// struct for passing precomputation data
+struct gaus_interp_params {
+  std::vector<double> as;
+  std::vector<double> bs;
+  double sig2;
+};
+
+namespace internal {
+
+/*
+find the smallest difference between successive elements in a sorted vector
+*/
+template <typename Tx>
+double min_diff(int n, std::vector<Tx> const& xs) {
+  double dmin = value_of(xs[1]) - value_of(xs[0]);
+  for (int i = 1; i < n - 1; i++) {
+    if (value_of(xs[i + 1]) - value_of(xs[i]) < dmin) {
+      dmin = value_of(xs[i + 1]) - value_of(xs[i]);
+    }
+  }
+  return dmin;
+}
+
+}  // namespace internal
+
+/**
+ * Given a set of reference points \f$(xs_i, ys_i)\f$, create a mollifier
+ * that intersects the reference points. This function requires as input
+ * a struct created by the function gaus_interp_precomp. The algorithm
+ * used to create the mollifier is an iterative algorithm that works
+ * as follows. First a linear
+ * interpolation is created through the reference points. Then, the
+ * linear interpolation is convolved with a Gaussian whose width is
+ * proportional the smallest distance between successive points
+ * \f$xs_i\f$ and \f$xs_{i+1}\f$. Since the convolution is unlikely to
+ * intersect the reference points, the y-values of the reference points
+ * are shifted and the process repeats itself until the interpolation
+ * intersects all reference points.
+ *
+ * @param xs vector of independent variable of reference points
+ * @param ys vector of dependent variable of reference points
+ * @param params a struct created by gaus_interp_precomp that
+ * @param x the point at which to evaluate the interpolation
+ * @return value of the interpolation at x
+ */
+template <typename Tx>
+inline return_type_t<Tx> gaus_interp(std::vector<double> const& xs,
+                                     std::vector<double> const& ys,
+                                     gaus_interp_params const& params,
+                                     Tx const& x) {
+  // enforce that interpolation point is between smallest and largest
+  // reference point
+  static char const* function = "gaus_interp";
+  check_less_or_equal(function, "Interpolation point", x, xs.back());
+  check_greater_or_equal(function, "Interpolation point", x, xs.front());
+  check_ordered(function, "xs", xs);
+  check_not_nan(function, "xs", xs);
+  check_not_nan(function, "ys", ys);
+  check_not_nan(function, "x", x);
+
+  const double NSTDS = 10;
+  int n = xs.size();
+
+  // create copy of xs so that endpoints can be extended
+  std::vector<double> xs2 = xs;
+
+  // extend out first and last lines for convolution
+  double sig = std::sqrt(params.sig2);
+  xs2[0] = xs[0] - NSTDS * sig;
+  xs2[n - 1] = xs[n - 1] + NSTDS * sig;
```

Comment:
Should you also check that xs > 0 for 0 sized vectors?

Reply:
good point, I added check to make sure that xs is of length at least 2